File: lib/features/pages/data/announcement_service.dart
import 'package:clashkingapp/core/services/api_service.dart';
import 'package:clashkingapp/core/services/game_data_service.dart';
import 'package:clashkingapp/features/pages/models/app_announcement.dart';
import 'package:flutter/foundation.dart';

class AnnouncementArchivePage {
  const AnnouncementArchivePage({
    required this.items,
    required this.hasMore,
    required this.nextOffset,
  });

  final List<AppAnnouncement> items;
  final bool hasMore;
  final int nextOffset;
}

class AnnouncementService {
  AnnouncementService({ApiService? apiService})
    : _apiService = apiService ?? ApiService.shared;

  final ApiService _apiService;

  Future<AppAnnouncement?> getActiveAnnouncement() async {
    final announcements = await getActiveAnnouncements();
    return announcements.isEmpty ? null : announcements.first;
  }

  Future<List<AppAnnouncement>> getActiveAnnouncements() async {
    try {
      final locale = await GameDataService.resolvePreferredLocale();
      final target = announcementTarget(
        isWeb: kIsWeb,
        platform: defaultTargetPlatform,
      );
      final response = await _apiService.get(
        '/app/announcements/active?target=$target&locale=${Uri.encodeQueryComponent(locale.languageCode)}',
        requiresAuth: false,
      );
      final items = response['items'];
      if (items is List) {
        return items
            .whereType<Map<String, dynamic>>()
            .map(AppAnnouncement.fromJson)
            .where(
              (announcement) =>
                  announcement.title.isNotEmpty &&
                  announcement.subtitle.isNotEmpty,
            )
            .toList();
      }
      final item = response['item'];
      if (item is! Map<String, dynamic>) {
        return const [];
      }
      final announcement = AppAnnouncement.fromJson(item);
      if (announcement.title.isEmpty || announcement.subtitle.isEmpty) {
        return const [];
      }
      return [announcement];
    } catch (_) {
      return const [];
    }
  }

  Future<AppAnnouncement?> getAnnouncement(String id) async {
    if (id.trim().isEmpty) return null;
    try {
      final locale = await GameDataService.resolvePreferredLocale();
      final response = await _apiService.get(
        '/app/announcements/${Uri.encodeComponent(id)}?locale=${Uri.encodeQueryComponent(locale.languageCode)}',
        requiresAuth: false,
      );
      final item = response['item'];
      if (item is! Map<String, dynamic>) {
        return null;
      }
      final announcement = AppAnnouncement.fromJson(item);
      if (announcement.title.isEmpty || announcement.subtitle.isEmpty) {
        return null;
      }
      return announcement;
    } catch (_) {
      return null;
    }
  }

  Future<AnnouncementArchivePage> getPublishedPosts({
    int limit = 20,
    int offset = 0,
  }) async {
    final locale = await GameDataService.resolvePreferredLocale();
    final target = announcementTarget(
      isWeb: kIsWeb,
      platform: defaultTargetPlatform,
    );
    final response = await _apiService.get(
      '/app/posts?target=$target&limit=$limit&offset=$offset&locale=${Uri.encodeQueryComponent(locale.languageCode)}',
      requiresAuth: false,
    );
    final rawItems = response['items'];
    final items = rawItems is List
        ? rawItems
              .whereType<Map<String, dynamic>>()
              .map(AppAnnouncement.fromJson)
              .where(
                (announcement) =>
                    announcement.title.isNotEmpty &&
                    announcement.subtitle.isNotEmpty,
              )
              .toList()
        : <AppAnnouncement>[];
    return AnnouncementArchivePage(
      items: items,
      hasMore: response['has_more'] == true,
      nextOffset:
          (response['next_offset'] as num?)?.toInt() ?? offset + items.length,
    );
  }
}

String announcementTarget({
  required bool isWeb,
  required TargetPlatform platform,
}) {
  if (isWeb) return 'all';
  return switch (platform) {
    TargetPlatform.iOS => 'ios',
    TargetPlatform.android => 'android',
    _ => 'all',
  };
}

File: expo/src/features/home/data/announcement-service.ts
import {
  ActiveAnnouncementsEndpoint,
  AnnouncementEndpoint,
  PostsEndpoint,
} from '@clashking/api-contracts/expo';
import { Effect } from 'effect';

import type { ContractApiService } from '../../../core/api/contract-api';
import { isRecord } from '../../player/models/parsing';
import { AppAnnouncement } from './app-announcement';

export type AnnouncementTarget = 'ios' | 'android' | 'all';

export interface AnnouncementArchivePage {
  readonly items: readonly AppAnnouncement[];
  readonly hasMore: boolean;
  readonly nextOffset: number;
}

export function announcementTarget(platform: string): AnnouncementTarget {
  return platform === 'ios' ? 'ios' : platform === 'android' ? 'android' : 'all';
}

export class AnnouncementService {
  constructor(
    private readonly api: ContractApiService,
    private readonly platform: string,
    private readonly locale: () => string,
  ) {}

  async getActiveAnnouncement(): Promise<AppAnnouncement | null> {
    return (await this.getActiveAnnouncements())[0] ?? null;
  }

  async getActiveAnnouncements(): Promise<readonly AppAnnouncement[]> {
    try {
      const response = await Effect.runPromise(
        this.api.execute(ActiveAnnouncementsEndpoint, {
          path: {},
          query: {
            target: announcementTarget(this.platform),
            locale: this.languageCode(),
          },
          body: {},
        }),
      );
      return decodeAnnouncementCollection(response);
    } catch {
      return [];
    }
  }

  async getAnnouncement(id: string): Promise<AppAnnouncement | null> {
    const targetId = id.trim();
    if (!targetId) return null;
    try {
      const response = await Effect.runPromise(
        this.api.execute(AnnouncementEndpoint, {
          path: { announcementId: targetId },
          query: { locale: this.languageCode() },
          body: {},
        }),
      );
      return decodeAnnouncement(response.item ?? response);
    } catch {
      return null;
    }
  }

  async getPublishedPosts(limit = 20, offset = 0): Promise<AnnouncementArchivePage> {
    const response = await Effect.runPromise(
      this.api.execute(PostsEndpoint, {
        path: {},
        query: {
          target: announcementTarget(this.platform),
          limit,
          offset,
          locale: this.languageCode(),
        },
        body: {},
      }),
    );
    const items = Array.isArray(response.items)
      ? response.items
          .map(decodeAnnouncement)
          .filter((item): item is AppAnnouncement => item !== null)
      : [];
    return {
      items,
      hasMore: response.has_more === true,
      nextOffset:
        typeof response.next_offset === 'number'
          ? Math.trunc(response.next_offset)
          : offset + items.length,
    };
  }

  private languageCode(): string {
    return this.locale().replace('-', '_').split('_', 1)[0] || 'en';
  }
}

function decodeAnnouncementCollection(response: Record<string, unknown>): AppAnnouncement[] {
  if (Array.isArray(response.items)) {
    return response.items
      .map(decodeAnnouncement)
      .filter((item): item is AppAnnouncement => item !== null);
  }
  const item = decodeAnnouncement(response.item);
  return item ? [item] : [];
}

function decodeAnnouncement(value: unknown): AppAnnouncement | null {
  if (!isRecord(value)) return null;
  const item = AppAnnouncement.fromJson(value);
  return item.title && item.subtitle ? item : null;
}
